src/lazy/encoder/binary/v1_1/flex_uint.rs

```diff
@@ -46,79 +45,49 @@ impl FlexUInt {
     ///          an appropriate error message if reading fails.
     #[inline(always)]
     pub fn read(input: &[u8], offset: usize) -> IonResult<FlexUInt> {
-        const COMMON_CASE_INPUT_BYTES_NEEDED: usize = 8;
-
+        const COMMON_CASE_INPUT_BYTES_NEEDED: usize = 4;
         // We want to minimize the number of branches that happen in the common case. To do this,
         // we perform a single length check, making sure that the buffer contains enough data to
         // represent a FlexUInt whose continuation bits fit in a single byte (i.e. one with 7 or
-        // fewer bytes of magnitude). If the buffer doesn't have at least 8 bytes in it or the
-        // FlexUInt we find requires more than 8 bytes to represent, we'll fall back to the general
+        // fewer bytes of magnitude). If the buffer doesn't have at least 4 bytes in it or the
+        // FlexUInt we find requires more than 4 bytes to represent, we'll fall back to the general
         // case.
-        if input.len() < COMMON_CASE_INPUT_BYTES_NEEDED || input[0] == 0 {
-            // Calling `read_flex_primitive_as_uint_no_inline` keeps this method small enough that
-            // the code for the common case can be inlined.
-            return Self::read_flex_primitive_as_uint_no_inline(
-                input,
-                offset,
-                "reading a FlexUInt",
-                false,
-            );
+        if input.len() >= COMMON_CASE_INPUT_BYTES_NEEDED {
+            'common_case: {
+                let num_encoded_bytes = input[0].trailing_zeros() as usize + 1;
+                // By branching on particular values, we make the value of `num_encoded_bytes` in their
+                // corresponding arm `const`. This allows us to use `read_n_bytes` to optimize for those
+                // sizes.
+                let mut buffer = [0u8; size_of::<u64>()];
+                match num_encoded_bytes {
+                    1 => Self::read_n_bytes::<1>(input, &mut buffer),
+                    2 => Self::read_n_bytes::<2>(input, &mut buffer),
+                    3 => Self::read_n_bytes::<3>(input, &mut buffer),
+                    4 => Self::read_n_bytes::<4>(input, &mut buffer),
```

Comment:
🪧 For the most common sizes, we use a `const` helper function to do the reading. It's very slightly faster in Rust v1.86. However, the `nightly` version of the compiler is able to do magical things with it (presumably because it's using a newer LLVM version). Nightly can be ~10% faster than stable in some cases.